### ISSUE TITLE
Review and replace `useEffect` usage

### ISSUE THREAD

**Issue by Chris-Petty** [msupply-foundation/open-msupply#5237]
## The suggestion

The useEffect hook is a common source of bugs and performance issues when misused in react. I've seen a few places where it's been used for things other than synchronising state with some external system such as an API request. As all of our API work is managed through abstractions on top of `useQuery` hook, it appears suspicious that we have >250 calls of `useEffect`.

React has some good docs specifically on when not to use `useEffect` https://react.dev/learn/you-might-not-need-an-effect

At a glance a good amount do seem to be legit, but there are definitely some that should probably use a different approach!

## Example use case

https://github.com/msupply-foundation/open-msupply/blob/f895b95dc9d8c49fd5b8af01ed8726d5ad62506a/client/packages/requisitions/src/RequestRequisition/ListView/ProgramRequisitionOptions.tsx#L41-L55

To avoid unnecessary re-renders, prefer to use event handler functions rather than `useEffect`:

https://github.com/msupply-foundation/open-msupply/blob/01fbcea02ff9d486327262d146ec09590715337c/client/packages/requisitions/src/RequestRequisition/ListView/ProgramRequisitionOptions.tsx#L41-L56

Alternatively `useReducer` might be suitable in certain situations. The main point is to only use the `useEffect` hook if synchronising with an external system.

### Why should we invest time in this?

- unnecessary useEffects create unnecessary re-renders which slows the app down and introduces odd bugs.
- useEffect is asynchronous and typically creates code that is hard to follow, so things should become a bit simpler to follow logically.

### Are there any risks associated with this change?

Other than standard regressions, not much!

### How much effort is required?

There are 250+ `useEffect` calls in OMS. I don't know how many are proper usage or not, this issue is to work that out!

### Agreed Solution

During cool down period/refactor fortnight, systematically review useEffects. Make an issue if problematic? One challenge is distributing the load if multiple people look, and recording what was looked at... tempting to just a comment on each `useEffect` like `// Usage is legit!` haha.

**Comment by andreievg** [comment]
In next omSup all team session, let's look at some less then ideal use cases of this and start the conversation or at least make sure that everyone is aware of anti patterns (proper use of useEffect)

**Comment by andreievg** [comment]
Also add example use cases here, it will be and epic for reference
